### PR TITLE
Add dynamic windows desktops

### PR DESCRIPTION
**Pull request by probakowski**
This change adds new resource: dynamic windows desktop. 
The difference between this new type and current windows desktop is that it doesn't have host UUID - it is meant to be picked up by windows desktop services by matching labels similar to dynamic registration of apps and databases.

### CODE REVIEW

File: api/proto/teleport/legacy/types/types.proto [hunk 1]
```diff
@@ -5410,6 +5410,42 @@ message WindowsDesktopSpecV3 {
...
+  // not joined to an Active Directory domain.
+  bool NonAD = 4 [(gogoproto.jsontag) = "non_ad"];
+  // ScreenSize specifies the size of the screen to use for sessions
+  // on this host. In most cases this should be unspecified, in which
+  // case Teleport will fill the browser window.
+  Resolution ScreenSize = 5 [(gogoproto.jsontag) = "screen_size,omitempty"];
+}
+
+// DynamicWindowsDesktopFilter are filters to apply when searching for windows desktops.
+message DynamicWindowsDesktopFilter {
```
Comment (zmb3) [inline api/proto/teleport/legacy/types/types.proto:5444]: Do we need to duplicate this? Can we use the existing filter message and just ignore the HostID?

File: api/types/desktop.go [hunk 2]
```diff
@@ -127,6 +127,198 @@ func (s *WindowsDesktopServiceV3) MatchSearch(values []string) bool {
...
+		return trace.Wrap(err)
+	}
+
+	if d.Spec.ScreenSize != nil {
+		if d.Spec.ScreenSize.Width > MaxRDPScreenWidth || d.Spec.ScreenSize.Height > MaxRDPScreenHeight {
+			return trace.BadParameter("invalid screen size %dx%d (maximum %dx%d)",
+				d.Spec.ScreenSize.Width, d.Spec.ScreenSize.Height, MaxRDPScreenWidth, MaxRDPScreenHeight)
+		}
+	}
+
```
Comment (zmb3) [inline api/types/desktop.go:199]: This is a lot of copy-paste from the original desktop resource. I'm worried about the two diverging. Any suggestions for sharing more code?

File: api/types/desktop.go [hunk 3]
```diff
@@ -127,6 +127,198 @@ func (s *WindowsDesktopServiceV3) MatchSearch(values []string) bool {
...
+func (s DynamicWindowsDesktops) AsResources() []ResourceWithLabels {
+	resources := make([]ResourceWithLabels, 0, len(s))
+	for _, server := range s {
+		resources = append(resources, ResourceWithLabels(server))
+	}
+	return resources
+}
+
+// GetFieldVals returns list of select field values.
+func (s DynamicWindowsDesktops) GetFieldVals(field string) ([]string, error) {
```
Comment (zmb3) [inline api/types/desktop.go:304]: What is this used for?

File: constants.go [hunk 4]
```diff
@@ -262,6 +262,9 @@ const (
 	// ComponentWindowsDesktop is a Windows desktop access server.
 	ComponentWindowsDesktop = "windows_desktop"
 
+	// ComponentDynamicWindowsDesktop is a Windows desktop access server.
```
Comment (zmb3) [inline constants.go:265]: This comment is exactly the same as the one on `ComponentWindowsDesktop`. I would expect the godoc to help me understand what is different about a dynamic desktop.

Then again, if this is just used for logging, then you can probably use the original component field.

File: lib/config/fileconf.go [hunk 5]
```diff
@@ -2385,7 +2385,8 @@ type WindowsDesktopService struct {
 	// HostLabels optionally applies labels to Windows hosts for RBAC.
 	// A host can match multiple rules and will get a union of all
 	// the matched labels.
-	HostLabels []WindowsHostLabelRule `yaml:"host_labels,omitempty"`
+	HostLabels       []WindowsHostLabelRule `yaml:"host_labels,omitempty"`
+	ResourceMatchers []ResourceMatcher      `yaml:"resources,omitempty"`
```
Comment (zmb3) [inline lib/config/fileconf.go:2389]: Add godoc for the new field?

**Review by zmb3 — COMMENTED** [review]
This is a pretty big PR. It might make sense to do the proto changes in a dedicated. PR.

**Review by rosstimothy — COMMENTED** [review]
Agree with Zac, there is a lot going on here. It might be best to split this into PRs that operate on a single layer: proto, backend storage, rpc implementation, tctl, resource watcher, windows desktop server changes.

**Comment by probakowski** [comment]
It was splitted to smaller PRs, closing this one

**Pull request closed without merging**